### ISSUE TITLE
No logs when deis used with k8s

### ISSUE THREAD

**Issue by Grindizer** [deis/deis#4838]
Hi all, 

We have a setup with a stateless deis 1.12.2 using k8s schedular, for now  we used the in memory logger, but the problem is we do not have any logs with deis logs command:

```shell
# deis logs -a alfred

<< empty response
```

we do have log when we use kubectl though and on the instance level docker log works as well, 

```shell
kubectl --namespace=alfred logs alfred-v2-cmd-33h04
Starting server... done.
time="2016-01-05T22:50:47Z" level=info msg="(377662431b14f424e58a14dde72fdd46) getImage" time=25.104065198s
time="2016-01-05T22:50:47Z" level=info msg="(377662431b14f424e58a14dde72fdd46) hips.Process" time=25.104211844s
time="2016-01-05T22:51:39Z" level=info msg="(377662431b14f424e58a14dde72fdd46) getImage" time=9.979237052s
```

```shell
core@ip-10-0-101-124 ~ $ docker  logs a1e8e540cecc
Starting server... done.
time="2016-01-05T22:50:47Z" level=info msg="(377662431b14f424e58a14dde72fdd46) getImage" time=25.104065198s
time="2016-01-05T22:50:47Z" level=info msg="(377662431b14f424e58a14dde72fdd46) hips.Process" time=25.104211844s
time="2016-01-05T22:51:39Z" level=info msg="(377662431b14f424e58a14dde72fdd46) getImage" time=9.979237052s
```

I know there is some work going on with logging on deis 2.0, but I was wondering what's the status of logging for 1.12.2 with k8s ?

**Comment by krancour** [comment]
This should be working.  Can you look at the values of the following two etcd keys and then check whether they seem to accurately point to the logger?

* `/deis/logs/host`
* `/deis/logs/port`

Generally speaking, logger publishes those and then the logspout on each node looks at those at startup to discover the logger.  In the past, people running stateless have sometimes overridden those values to bypass the logger and send logs straight from logspout to some external log archive.  (There's no longer any good reason to bypass the logger now that the in-memory log storage is available.  Logger can drain to an external archive instead if that's still desired.)

The salient point is... check that those etcd keys accurately point the logspout instances at where the logger is _actually running_.

One thing to note is issue #4293.  Because of that, logspout instances really need to be restarted if/when logger is rescheduled from one node to another.

Apart from all this, the only other possibility is that logspout isn't running or has "seized up" for some reason-- an issue we've seen in the past that we believe has been corrected.

**Comment by Grindizer** [comment]
Thanks for the answer, here for the etcd keys, 
```shell
core@ip-10-0-101-124 ~ $ etcdctl get /deis/logs/host
10.0.140.126
core@ip-10-0-101-124 ~ $ etcdctl get /deis/logs/port
514
```

It seems that the host is correct: 10.0.140.126
```shell
core@ip-10-0-101-124 ~ $ fleetctl list-units
UNIT					MACHINE				ACTIVE	SUB
deis-builder.service			27cf4e1e.../10.0.140.126	active	running
deis-controller.service			374e5749.../10.0.118.5		active	running
deis-kube-apiserver.service		27cf4e1e.../10.0.140.126	active	running
deis-kube-controller-manager.service	27cf4e1e.../10.0.140.126	active	running
deis-kube-kubelet.service		bea10648.../10.0.101.124	active	running
deis-kube-kubelet.service		de6ac0cb.../10.0.148.181	active	running
deis-kube-proxy.service			50416cee.../10.0.116.129	active	running
deis-kube-proxy.service			665c0c4b.../10.0.190.216	active	running
deis-kube-scheduler.service		27cf4e1e.../10.0.140.126	active	running
deis-logger.service			27cf4e1e.../10.0.140.126	active	running
deis-logspout.service			27cf4e1e.../10.0.140.126	active	running
deis-logspout.service			374e5749.../10.0.118.5		active	running
deis-logspout.service			50416cee.../10.0.116.129	active	running
deis-logspout.service			665c0c4b.../10.0.190.216	active	running
deis-logspout.service			bea10648.../10.0.101.124	active	running
deis-logspout.service			de6ac0cb.../10.0.148.181	active	running
deis-publisher.service			bea10648.../10.0.101.124	active	running
deis-publisher.service			de6ac0cb.../10.0.148.181	active	running
deis-registry@1.service			374e5749.../10.0.118.5		active	running
deis-router@1.service			50416cee.../10.0.116.129	active	running
deis-router@2.service			665c0c4b.../10.0.190.216	active	running
```

The port is correct as well: 514
```shell
core@ip-10-0-140-126 ~ $ docker ps
CONTAINER ID        IMAGE                   COMMAND                CREATED             STATUS              PORTS                                                   NAMES
dcde37b27106        deis/builder:v1.12.1    "/bin/entry /bin/boo   2 days ago          Up 2 days           0.0.0.0:2223->2223/tcp                                  deis-builder
f304d7aef77e        deis/logspout:v1.12.1   "/bin/logspout"        2 days ago          Up 2 days                                                                   deis-logspout
7e3f01a4ceae        deis/logger:v1.12.1     "/bin/logger --enabl   2 days ago          Up 2 days           0.0.0.0:514->514/udp, 514/tcp, 0.0.0.0:8088->8088/tcp   deis-logger
a1226c6b8978        sysdig/agent            "/docker-entrypoint.   2 days ago          Up 2 days                                                                   sysdig-agent
```

Logger seems to be up, but has no data regarding my app (alfred)
```shell
core@ip-10-0-140-126 ~ $ docker logs --tail=all deis-logger
2016/01/05 20:57:32 configurer running
2016/01/05 20:57:42 configurer: Activated new storage adapter: memory:2000
2016/01/05 20:57:43 syslogish server running
2016/01/05 20:57:43 weblog server running
2016/01/05 20:57:43 publisher running
2016/01/05 20:57:43 deis-logger running
2016/01/07 18:28:19 Could not find logs for 'alfred'
2016/01/07 18:28:20 Could not find logs for 'alfred'
```

The logsprout on the instance containing the pod seems to be running.
```shell
core@ip-10-0-101-124 ~ $ docker logs --tail=all deis-logspout
2016/01/07 21:25:53 attach: e3e2f607a78a deis-logspout success
2016/01/07 21:25:53 attach: a1e8e540cecc k8s_alfred-cmd.a21bfbc0_alfred-v2-cmd-33h04_alfred_7f0140d1-b3fe-11e5-b730-0a4263846359_21f5274f success
2016/01/07 21:25:53 attach: ae2b0847b46e k8s_POD.e4cc795_alfred-v2-cmd-33h04_alfred_7f0140d1-b3fe-11e5-b730-0a4263846359_9806dce9 success
2016/01/07 21:25:53 attach: 3a199a5d1638 deis-publisher success
2016/01/07 21:25:53 attach: bb7e0364fcf6 sysdig-agent success
2016/01/07 21:25:53 etcd: http://10.0.101.124:4001
2016/01/07 21:25:53 routing all to udp://10.0.140.126:514
2016/01/07 21:25:53 loading and persisting routes in /tmp
2016/01/07 21:25:53 logspout serving http on :8000
```

I don't know how logs should looks like when logsrpout send data to logger, 
but it seems there is something missing here, either data are not sent or logger does not store the data ? 

any idea would be more then welcome, and thanks again for your help.

**Comment by krancour** [comment]
Use `docker ps` on the node hosting alfred to find the container id.  Then you can grep the logspout logs on that same host to see if logspout has "attached" to that container.

**Comment by Grindizer** [comment]
Ok. here with another application:

here just to check app logs, and get node on which a pod of my app is running:
```shell
(hearst)nassims-MacBook-Pro:devops nacim$ deis logs -a rover

(hearst)nassims-MacBook-Pro:devops nacim$ kubectl --namespace=rover get pods -o wide
NAME                 READY     STATUS    RESTARTS   AGE       NODE
rover-v8-cmd-z6wzs   1/1       Running   0          11m       10.0.135.161
```
I ssh to the node:
```shell
(hearst)nassims-MacBook-Pro:devops nacim$ ssh 10.0.135.161 -F provisioning/envs/devops/ssh.config
Warning: Permanently added '10.0.135.161' (RSA) to the list of known hosts.
Last login: Sun Jan 10 17:59:01 2016 from 10.0.23.45
CoreOS stable (766.4.0)
Update Strategy: No Reboots
Failed Units: 1
  update-engine.service
core@ip-10-0-135-161 ~ $ docker ps
CONTAINER ID        IMAGE                                  COMMAND                CREATED             STATUS              PORTS               NAMES
48db454fc867        10.0.167.120:5000/rover:v8             "./bin/boot"           11 minutes ago      Up 11 minutes                           k8s_rover-cmd.9005fbc7_rover-v8-cmd-z6wzs_rover_82f14b08-b7c2-11e5-946e-12d891e80f69_3919bd5f
8a607b8367c4        gcr.io/google_containers/pause:0.8.0   "/pause"               11 minutes ago      Up 11 minutes                           k8s_POD.e4cc795_rover-v8-cmd-z6wzs_rover_82f14b08-b7c2-11e5-946e-12d891e80f69_9fa41800
f63811bb51a6        10.0.167.120:5000/alfred:v2            "/go/bin/alfred"       18 hours ago        Up 18 hours                             k8s_alfred-cmd.d533fc33_alfred-v2-cmd-c4z55_alfred_da12cd0d-b727-11e5-946e-12d891e80f69_1925fec2
959ef0761ff2        gcr.io/google_containers/pause:0.8.0   "/pause"               18 hours ago        Up 18 hours                             k8s_POD.e4cc795_alfred-v2-cmd-c4z55_alfred_da12cd0d-b727-11e5-946e-12d891e80f69_db1f8733
be8efcac2133        deis/logspout:v1.12.1                  "/bin/logspout"        42 hours ago        Up 42 hours                             deis-logspout
de91b56b1db7        deis/publisher:v1.12.1                 "/usr/local/bin/publ   46 hours ago        Up 46 hours                             deis-publisher
7204d2b0a937        sysdig/agent                           "/docker-entrypoint.   47 hours ago        Up 47 hours                             sysdig-agent
```

and here I get logsprout logs, I can see something like "attach" for my container.
```shell
core@ip-10-0-135-161 ~ $ docker logs be8efcac2133 &> /tmp/logs && cat /tmp/logs | grep 48db454fc867
2016/01/10 17:49:42 event: 48db454fc867 create
2016/01/10 17:49:42 event: 48db454fc867 start
2016/01/10 17:49:42 attach: 48db454fc867 k8s_rover-cmd.9005fbc7_rover-v8-cmd-z6wzs_rover_82f14b08-b7c2-11e5-946e-12d891e80f69_3919bd5f success
2016/01/10 17:50:04 event: 48db454fc867 exec_create: env
2016/01/10 17:50:04 event: 48db454fc867 exec_start: env
core@ip-10-0-135-161 ~ $
```

and here I check if docker logs on my container actually contains data:
```shell
core@ip-10-0-135-161 ~ $ docker logs 48db454fc867
/app/rover/urls.py:87: RemovedInDjango110Warning: Support for string view arguments to url() is deprecated and will be removed in Django 1.10 (got content.views.healthcheck). Pass the callable instead.
  url(r'^healthcheck/?$', 'content.views.healthcheck'),

System check identified some issues:

WARNINGS:
?: (urls.W002) Your URL pattern '^/?$' has a regex beginning with a '/'. Remove this slash as it is unnecessary.
Operations to perform:
  Apply all migrations: content, contenttypes, auth
Running migrations:
  No migrations to apply.
[2016-01-10 17:49:43 +0000] [11] [INFO] Starting gunicorn 19.4.1
[2016-01-10 17:49:43 +0000] [11] [INFO] Listening at: http://0.0.0.0:80 (11)
[2016-01-10 17:49:43 +0000] [11] [INFO] Using worker: sync
[2016-01-10 17:49:43 +0000] [14] [INFO] Booting worker with pid: 14
core@ip-10-0-135-161 ~ $
```

So to be honest, I suspect logsprout not being reading the logs from the right source of data ? could that be linked to the fact we use k8s schedular ?

**Comment by bacongobbler** [comment]
Just a quick FYI that the kubernetes scheduler in v1 is a technical preview and is not to be used in production (see the note [here](http://docs.deis.io/en/latest/customizing_deis/choosing-a-scheduler/#kubernetes-scheduler)). It's just a proof of concept and there are likely going to be bugs.

That being said, this is likely an issue with the kubernetes scheduler as the pod names are distinctly different than what a fleet unit name looks like, though I'd have to reproduce the issue to get a better view of the problem.

**Comment by krancour** [comment]
@Grindizer, I'm at a loss in terms of what's going on here.  @bacongobbler mentioned the difference in how things are named being a potential culprit.  I've checked up on those regexes and decided to double-check them.

TL;DR-- nothing wrong with the container names or the regular expressions that try to match them.

With a container named `k8s_rover-cmd.9005fbc7_rover-v8-cmd-z6wzs_rover_82f14b08-b7c2-11e5-946e-12d891e80f69_3919bd5f`, this line in the logspout code:

https://github.com/deis/deis/blob/master/logspout/logspout.go#L109

Should match it and extract the tokens:

* rover
* cmd-z6wzs

This line incorporates those into a log message:

https://github.com/deis/deis/blob/master/logspout/logspout.go#L87-L92

The format of that message will be:

```
<timestamp> rover[cmd-z6wzs]: <message>
```

On the logger end, the app name is extracted from the message using this regex:

https://github.com/deis/deis/blob/master/logger/syslogish/server.go#L21

Given a message formatted as indicated above and given that regular expression, it seems that `rover` is correctly identified as the application that the log message originated from.  It _should_ be slotted into the ringbuffer for that application.

fwiw, I use this little online tool for testing regular expressions:

https://regex-golang.appspot.com/assets/html/index.html

**Comment by krancour** [comment]
The only remaining thing I can think of is to ask you to check the value of the `/deis/logs/storageAdapterType` key in etcd.

**Comment by DanHoerst** [comment]
@krancour 

    core@ip-10-x ~ $ etcdctl get /deis/logs/storageAdapterType
    memory:2000

**Comment by DanHoerst** [comment]
@krancour It looks like this was our fault. After your last response about how it "should" work, I went back over everything and noticed we didn't have the correct udp ports open to allow the syslog traffic back and forth - we had tcp instead. Sorry to waste your time but thanks for the clear explanations and quick responses.

**Comment by krancour** [comment]
I've made that same mistake before, so don't feel too bad.

@Grindizer that may be something for you to check as well.

**Comment by DanHoerst** [comment]
@krancour I'm working with @Grindizer . It was the same issue.

**Comment by krancour** [comment]
I suspected, but didn't know that for sure.  Thanks.

**Comment by Grindizer** [comment]
@krancour really thank you again for your time, if it can help we at least learn better how to debug logging in Deis (I know, it's a kinda thin :o)).

**Comment by krancour** [comment]
It was no trouble.  You won't be the only ones to encounter this difficulty.  The actual root cause and all of the hypothesizing as to other possible causes will certainly help  others in the community eventually, and I'm sure the greater insight you now have into Deis will pay dividends as well.  Pay it forward.  :)
